### PR TITLE
Functional and Performance tests

### LINKED ISSUE #49: Make a benchmarks
Compare to Pomelo.EntityFrameworkCore.MySql-1.0.0(rtm), Pomelo.EntityFrameworkCore.MySql-1.0.1(prerelease), MySql.Data.EntityFrameworkCore(7.0.5)

### PR DESCRIPTION
**Pull request by caleblloyd**
This PR adds a test application that implements both Functional and Performance tests.  Instructions on running Functional and Performance tests can be found at [test/Pomelo.EntityFrameworkCore.MySql.Functional/README.md](https://github.com/caleblloyd/Pomelo.EntityFrameworkCore.MySql/tree/f_functional_tests/test/Pomelo.EntityFrameworkCore.MySql.Functional/README.md)

The Functional tests uses MSTest and are pretty much a shell right now.  I wrote 1 example test to show how it works.  More functional tests should be added.

I've run the performance tests using [MySqlConnector/0.1.0-alpha21](https://www.nuget.org/packages/MySqlConnector/0.1.0-alpha21).  We are running at about 100RPS on Async Targets right now before hitting issues:

	Rate = 100 RPS
	Duration = 5s
	Targets = targets-async.txt
	Requests      [total, rate]            500, 100.20
	Duration      [total, attack, wait]    5.007885059s, 4.989999936s, 17.885123ms
	Latencies     [mean, 50, 95, 99, max]  18.269278ms, 13.847945ms, 49.695278ms, 77.360226ms, 108.858535ms
	Bytes In      [total, mean]            13798561, 27597.12
	Bytes Out     [total, mean]            360012, 720.02
	Success       [ratio]                  100.00%
	Status Codes  [code:count]             200:500 

After 100RPS we begin to get lock-ups.  I suspect this has something to do with Pomelo locking a thread where it shouldn't be.  MySqlConnector/0.1.0-alpha21 is known to support up to 1500RPS easily on the hardware I'm using.

My next focus will be reducing these performance bottlenecks and achieving as close to 1500RPS as possible with Pomelo.  I'll also be working on Functional tests that ensure the proper data types are being created in MySQL.

Let's merge this PR so others can start to contribute to the Functional tests and test the performance impacts of their changes!

Fixed #49

### CODE REVIEW

**Comment by caleblloyd** [comment]
@Kagamine it looks like AppVeyor is picking up that there is a new test project, which requires a database.  Long term, we probably want AppVeyor to run these functional tests.  Any way to add a MySql database to AppVeyor?

**Comment by Kagamine** [comment]
Performance tests looks great but we still need a FunctionalTests which should be forked from https://github.com/aspnet/EntityFramework/tree/dev/test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests.

**Comment by Kagamine** [comment]
The default pwd is `Password12!`

See https://www.appveyor.com/docs/services-databases/#mysql

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/config.json.example [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿{
+  "Data": {
+    "ConnectionString": "server=127.0.0.1;user id=root;password=abc123;port=3306;database=pomelo_test;"
```
Comment (Kagamine) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/config.json.example]: https://www.appveyor.com/docs/services-databases/#mysql

The default pwd of mysql in appveyor is `Password12!`

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/Tests/Models/DataTypesTest.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                    TypeUshort     = 2,
+                    TypeInt        = 3,
+                    TypeUint       = 4,
+                    TypeLong       = 5,
+                    TypeUlong      = 6,
+                    TypeString     = "test",
+                    TypeByte       = (byte) 'a',
+                    TypeByteArray  = new byte[] { (byte) 'a', (byte) 'b' },
+                    TypeJsonObj    = new JsonObject<List<string>>(new List<string>{ "test" }),     
+                    TypeJsonArray  = new JsonObject<Dictionary<string, string>>(new Dictionary<string, string>(){{ "test", "test" }})
```
Comment (Kagamine) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/Tests/Models/DataTypesTest.cs]: Reverse the names of TypeJsonArray and TypeJsonObj. Besides, I think `TypeJsonObject` is better than `TypeJsonObj`

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/project.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
+﻿{
+  "dependencies": {
+    "dotnet-test-mstest": "1.1.1-preview",
+    "MSTest.TestFramework": "1.0.1-preview",
```
Comment (Kagamine) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/project.json]: Could you switch the test framework into xUnit?

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/Models/DataTypes.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        // string
+        public string   TypeString    { get; set; }
+        
+        // binary
+        public byte     TypeByte      { get; set; }
+        public byte[]   TypeByteArray { get; set; }
+
+        // json
+        public JsonObject<List<string>>               TypeJsonObj    { get; set; }
+        public JsonObject<Dictionary<string, string>> TypeJsonArray  { get; set; }
```
Comment (Kagamine) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/Models/DataTypes.cs]: Reverse the names of TypeJsonObj & TypeJsonArray

File: .vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (Kagamine) [inline .vscode/launch.json]: Add `/.vscode` into gitignore.

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/Program.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.AspNetCore.Hosting;
+
+namespace Pomelo.EntityFrameworkCore.MySql.Functional
```
Comment (Kagamine) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/Program.cs]: Should `Pomelo.EntityFrameworkCore.MySql.PerfTests` be better?

**Comment by Kagamine** [comment]
I still want to compare our lib to the official one.

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/project.json (hunk 3 above)
Comment (caleblloyd) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/project.json]: I will try it!  Shouldn't be an issue if XUnit still works with Dependency Injection

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/Models/DataTypes.cs (hunk 4 above)
Comment (caleblloyd) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/Models/DataTypes.cs]: Haha oops.  good catch

File: test/Pomelo.EntityFrameworkCore.MySql.Functional/Program.cs (hunk 6 above)
Comment (caleblloyd) [inline test/Pomelo.EntityFrameworkCore.MySql.Functional/Program.cs]: Yes, I'll change it

**Comment by caleblloyd** [comment]
> I still want to compare our lib to the official one.

@Kagamine I have been doing this.  Oracle has a bug with their JOIN requests, so I can't performance test GETs.  I have been able to performance test Inserts via POSTs though.

**Oracle is hitting around 600 RPS steady with Inserts Only:**

	Pomelo.EntityFrameworkCore.MySql.PerfTests$ export EF_PROVIDER="oracle"
	Pomelo.EntityFrameworkCore.MySql.PerfTests$ dotnet run
	Pomelo.EntityFrameworkCore.MySql.PerfTests/scripts$ ./stress.sh 600 10s
	Rate = 600 RPS
	Duration = 10s
	Targets = targets-async.txt
	Requests      [total, rate]            6000, 600.10
	Duration      [total, attack, wait]    10.301519169s, 9.998328674s, 303.190495ms
	Latencies     [mean, 50, 95, 99, max]  153.350372ms, 123.880774ms, 312.361365ms, 325.139826ms, 338.753755ms
	Bytes In      [total, mean]            21720000, 3620.00
	Bytes Out     [total, mean]            21444000, 3574.00
	Success       [ratio]                  100.00%
	Status Codes  [code:count]             200:6000  

**We have a slowdown somewhere in our Library. Pomelo runs fine up to 200 RPS (inserts only)**

	Pomelo.EntityFrameworkCore.MySql.PerfTests$ export EF_PROVIDER="pomelo"
	Pomelo.EntityFrameworkCore.MySql.PerfTests$ dotnet run
	Pomelo.EntityFrameworkCore.MySql.PerfTests/scripts$ ./stress.sh 200 10s
	Rate = 200 RPS
	Duration = 10s
	Targets = targets-async.txt
	Requests      [total, rate]            2000, 200.10
	Duration      [total, attack, wait]    9.99961143s, 9.994999771s, 4.611659ms
	Latencies     [mean, 50, 95, 99, max]  4.755548ms, 4.186371ms, 7.434174ms, 14.878961ms, 33.346662ms
	Bytes In      [total, mean]            7240000, 3620.00
	Bytes Out     [total, mean]            7148000, 3574.00
	Success       [ratio]                  100.00%
	Status Codes  [code:count]             200:2000 

**Pomelo slowdown and lockup at 250RPS:**

	./stress.sh 250 10s
	Rate = 250 RPS
	Duration = 10s
	Targets = targets-async.txt
	Requests      [total, rate]            2500, 250.10
	Duration      [total, attack, wait]    39.998119025s, 9.995999706s, 30.002119319s
	Latencies     [mean, 50, 95, 99, max]  1.588906722s, 4.132237ms, 30.000919664s, 30.002019899s, 30.020246016s
	Bytes In      [total, mean]            8572160, 3428.86
	Bytes Out     [total, mean]            8466806, 3386.72
	Success       [ratio]                  94.72%
	Status Codes  [code:count]             200:2368  500:1  0:131  

 I'm working to find it, will spend more time on it tonight.

**Comment by caleblloyd** [comment]
All of the code review comments have been fixed in latest commit.  Switched namespace and switched to XUnit.  I am now working to get CI up

**Comment by caleblloyd** [comment]
@Kagamine fixed up the CI configs, we are :white_check_mark:

**Comment by Kagamine** [comment]
:shipit:

**Pull request merged**